guppylang/std/_internal/compiler/prelude.py

```diff
@@ -151,3 +158,36 @@ def compile_with_inouts(self, args: list[Wire]) -> CallReturnWires:
 
     def compile(self, args: list[Wire]) -> list[Wire]:
         raise InternalGuppyError("Call compile_with_inouts instead")
+
+
+class UnwrapOpCompiler(CustomInoutCallCompiler):
+    """Compiler for operations that require unwrapping an single option value result
+    which can potentially cause an error.
+
+    Args:
+        op: A HUGR operation that outputs an option value.
+        err_msg: The error message to be raised if the unwrapped option is an error.
+    """
+
+    op: Callable[[ht.FunctionType, Inst], ops.DataflowOp]
+    err_msg: str
+
+    def __init__(self, op: Callable[[ht.FunctionType, Inst], ops.DataflowOp], err: str):
+        self.op = op
+        self.err_msg = err
+
+    def compile_with_inouts(self, args: list[Wire]) -> CallReturnWires:
+        assert len(self.ty.output) == 1
+        # To instantiate the op we need a function signature that wraps the output of
+        # the function that is being compiled into a sum type.
+        opt_func_type = ht.FunctionType(
+            input=self.ty.input,
+            output=[ht.Either([error_type()], self.ty.output)],
+        )
+        op = self.op(opt_func_type, self.type_args)
+        option = self.builder.add_op(op, *args)
+        result = build_unwrap(self.builder, option, self.err_msg)
```

Comment:
If the op alreday returns an `error_type()`, imo it would be better to panic with that value instead of making up a new message

Reply:
Is there a good way in general to get the value of a wire - so I know how to check if this is an `error_type` but not how to get the `ErrorVal` which has the message?

Reply:
`Option` is a wire with type `Either<error, T>`. Just build a conditional on that wire:
* In the left case, you'll get the error value and can panic with it
* In the right case, you get the unwrapped value and can return it

You won't be able to rely on `build_unwrap` for that since that was only designed for `Option`, not `Either<error, T>`.

Reply:
Maybe we should call it `unwrap_result`?